File: lib/services/data_service.dart

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';
import '../models/transaction.dart';
import '../models/category.dart';

class DataService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  
  // Collection references
  CollectionReference get _transactionsCollection => _firestore.collection('transactions');
  CollectionReference get _categoriesCollection => _firestore.collection('categories');

  // Initial categories
  static final List<Category> _initialCategories = [
    Category(
      id: "cat1",
      name: "Salaire",
      icon: "account_balance",
      color: Colors.green,
      type: CategoryType.income,
    ),
    Category(
      id: "cat2",
      name: "Investissements",
      icon: "trending_up",
      color: Colors.blue,
      type: CategoryType.income,
    ),
    Category(
      id: "cat3",
      name: "Courses",
      icon: "shopping_cart",
      color: Colors.orange,
      type: CategoryType.expense,
    ),
    Category(
      id: "cat4",
      name: "Transport",
      icon: "directions_car",
      color: Colors.blueGrey,
      type: CategoryType.expense,
    ),
    Category(
      id: "cat5",
      name: "Divertissement",
      icon: "tv",
      color: Colors.purple,
      type: CategoryType.expense,
    ),
    Category(
      id: "cat6",
      name: "Restauration",
      icon: "restaurant",
      color: Colors.pink,
      type: CategoryType.expense,
    ),
  ];

  // Initial transactions
  static final List<Transaction> _initialTransactions = [
    Transaction(
      id: "t1",
      amount: 3000,
      type: TransactionType.income,
      categoryId: "cat1",
      description: "Salaire mensuel",
      date: DateTime(2025, 4, 1),
    ),
    Transaction(
      id: "t2",
      amount: 500,
      type: TransactionType.income,
      categoryId: "cat2",
      description: "Dividendes",
      date: DateTime(2025, 4, 5),
    ),
    Transaction(
      id: "t3",
      amount: 150,
      type: TransactionType.expense,
      categoryId: "cat3",
      description: "Courses hebdomadaires",
      date: DateTime(2025, 4, 7),
    ),
    Transaction(
      id: "t4",
      amount: 45,
      type: TransactionType.expense,
      categoryId: "cat4",
      description: "Essence",
      date: DateTime(2025, 4, 8),
    ),
    Transaction(
      id: "t5",
      amount: 30,
      type: TransactionType.expense,
      categoryId: "cat5",
      description: "Billets de cinéma",
      date: DateTime(2025, 4, 9),
    ),
    Transaction(
      id: "t6",
      amount: 75,
      type: TransactionType.expense,
      categoryId: "cat6",
      description: "Dîner avec des amis",
      date: DateTime(2025, 4, 10),
    ),
  ];

  // Initialize Firestore with default data if empty
  Future<void> initializeFirestoreData() async {
    // Check if categories collection is empty
    final categoriesSnapshot = await _categoriesCollection.get();
    if (categoriesSnapshot.docs.isEmpty) {
      // Add initial categories
      for (var category in _initialCategories) {
        await _categoriesCollection.doc(category.id).set(category.toJson());
      }
    }
    
    // Check if transactions collection is empty
    final transactionsSnapshot = await _transactionsCollection.get();
    if (transactionsSnapshot.docs.isEmpty) {
      // Add initial transactions
      for (var transaction in _initialTransactions) {
        await _transactionsCollection.doc(transaction.id).set(transaction.toJson());
      }
    }
  }

  // Get all transactions
  Future<List<Transaction>> getTransactions() async {
    await initializeFirestoreData();
    
    final snapshot = await _transactionsCollection.get();
    return snapshot.docs.map((doc) => 
      Transaction.fromJson(doc.data() as Map<String, dynamic>)
    ).toList();
  }

  // Get all categories
  Future<List<Category>> getCategories() async {
    await initializeFirestoreData();
    
    final snapshot = await _categoriesCollection.get();
    return snapshot.docs.map((doc) => 
      Category.fromJson(doc.data() as Map<String, dynamic>)
    ).toList();
  }

  // Add a new transaction
  Future<Transaction> addTransaction(Transaction transaction) async {
    await _transactionsCollection.doc(transaction.id).set(transaction.toJson());
    return transaction;
  }

  // Delete a transaction
  Future<void> deleteTransaction(String id) async {
    await _transactionsCollection.doc(id).delete();
  }

  // Add a new category
  Future<Category> addCategory(Category category) async {
    await _categoriesCollection.doc(category.id).set(category.toJson());
    return category;
  }

  // Update a category
  Future<Category> updateCategory(Category updatedCategory) async {
    await _categoriesCollection.doc(updatedCategory.id).set(updatedCategory.toJson());
    return updatedCategory;
  }

  // Delete a category
  Future<void> deleteCategory(String id) async {
    // Delete the category
    await _categoriesCollection.doc(id).delete();
    
    // Find and delete all transactions with this category
    final snapshot = await _transactionsCollection
        .where('categoryId', isEqualTo: id)
        .get();
    
    final batch = _firestore.batch();
    for (var doc in snapshot.docs) {
      batch.delete(doc.reference);
    }
    
    await batch.commit();
  }

  // Get transaction summary
  Future<Map<String, double>> getTransactionSummary() async {
    final transactions = await getTransactions();
    
    final double totalIncome = transactions
        .where((t) => t.type == TransactionType.income)
        .fold(0, (total, t) => total + t.amount);
        
    final double totalExpense = transactions
        .where((t) => t.type == TransactionType.expense)
        .fold(0, (total, t) => total + t.amount);
        
    final double balance = totalIncome - totalExpense;
    
    return {
      'totalIncome': totalIncome,
      'totalExpense': totalExpense,
      'balance': balance
    };
  }

  // Get category summary
  Future<List<Map<String, dynamic>>> getCategorySummary(TransactionType type) async {
    final transactions = await getTransactions();
    final categories = await getCategories();
    
    // Filter by transaction type
    final filteredTransactions = transactions.where((t) => t.type == type).toList();
    
    // Group by category
    final categorySummary = categories
        .where((c) => 
            c.type == (type == TransactionType.income 
              ? CategoryType.income 
              : CategoryType.expense) || 
            c.type == CategoryType.both)
        .map((category) {
          final categoryTransactions = filteredTransactions
              .where((t) => t.categoryId == category.id)
              .toList();
          
          final double total = categoryTransactions.fold(
              0, (sum, t) => sum + t.amount);
          
          return {
            'category': category,
            'total': total,
            'count': categoryTransactions.length
          };
        })
        .where((summary) => summary['count'] > 0)
        .toList();
    
    // Sort by total amount (descending)
    categorySummary.sort((a, b) => (b['total'] as double).compareTo(a['total'] as double));
        
    return categorySummary;
  }
}

File: src/services/dataService.ts

import { Transaction, Category } from "@/types";

// Mock data for categories
const initialCategories: Category[] = [
  {
    id: "cat1",
    name: "Salary",
    icon: "banknote",
    color: "#4CAF50",
    type: "income"
  },
  {
    id: "cat2",
    name: "Investments",
    icon: "trending-up",
    color: "#2196F3",
    type: "income"
  },
  {
    id: "cat3",
    name: "Groceries",
    icon: "shopping-cart",
    color: "#FF9800",
    type: "expense"
  },
  {
    id: "cat4",
    name: "Transport",
    icon: "car",
    color: "#607D8B",
    type: "expense"
  },
  {
    id: "cat5",
    name: "Entertainment",
    icon: "tv",
    color: "#9C27B0",
    type: "expense"
  },
  {
    id: "cat6",
    name: "Dining",
    icon: "utensils",
    color: "#E91E63",
    type: "expense"
  }
];

// Mock data for transactions
const initialTransactions: Transaction[] = [
  {
    id: "t1",
    amount: 3000,
    type: "income",
    categoryId: "cat1",
    description: "Monthly salary",
    date: "2025-04-01"
  },
  {
    id: "t2",
    amount: 500,
    type: "income",
    categoryId: "cat2",
    description: "Dividend payment",
    date: "2025-04-05"
  },
  {
    id: "t3",
    amount: 150,
    type: "expense",
    categoryId: "cat3",
    description: "Weekly groceries",
    date: "2025-04-07"
  },
  {
    id: "t4",
    amount: 45,
    type: "expense",
    categoryId: "cat4",
    description: "Fuel",
    date: "2025-04-08"
  },
  {
    id: "t5",
    amount: 30,
    type: "expense",
    categoryId: "cat5",
    description: "Movie tickets",
    date: "2025-04-09"
  },
  {
    id: "t6",
    amount: 75,
    type: "expense",
    categoryId: "cat6",
    description: "Dinner with friends",
    date: "2025-04-10"
  }
];

// Load data from localStorage if available
const loadData = <T>(key: string, initialData: T): T => {
  try {
    const storedData = localStorage.getItem(key);
    return storedData ? JSON.parse(storedData) : initialData;
  } catch (error) {
    console.error(`Error loading ${key} from localStorage:`, error);
    return initialData;
  }
};

// Save data to localStorage
const saveData = <T>(key: string, data: T): void => {
  try {
    localStorage.setItem(key, JSON.stringify(data));
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error);
  }
};

// Fetch all transactions
export const getTransactions = (): Transaction[] => {
  return loadData<Transaction[]>("transactions", initialTransactions);
};

// Fetch all categories
export const getCategories = (): Category[] => {
  return loadData<Category[]>("categories", initialCategories);
};

// Add a new transaction
export const addTransaction = (transaction: Omit<Transaction, "id">): Transaction => {
  const transactions = getTransactions();
  const newTransaction = {
    ...transaction,
    id: `t${Date.now()}`
  };
  
  const updatedTransactions = [...transactions, newTransaction];
  saveData("transactions", updatedTransactions);
  
  return newTransaction;
};

// Delete a transaction
export const deleteTransaction = (id: string): void => {
  const transactions = getTransactions();
  const updatedTransactions = transactions.filter(t => t.id !== id);
  saveData("transactions", updatedTransactions);
};

// Add a new category
export const addCategory = (category: Omit<Category, "id">): Category => {
  const categories = getCategories();
  const newCategory = {
    ...category,
    id: `cat${Date.now()}`
  };
  
  const updatedCategories = [...categories, newCategory];
  saveData("categories", updatedCategories);
  
  return newCategory;
};

// Update a category
export const updateCategory = (updatedCategory: Category): Category => {
  const categories = getCategories();
  const updatedCategories = categories.map(category => 
    category.id === updatedCategory.id ? updatedCategory : category
  );
  
  saveData("categories", updatedCategories);
  return updatedCategory;
};

// Delete a category
export const deleteCategory = (id: string): void => {
  const categories = getCategories();
  const updatedCategories = categories.filter(c => c.id !== id);
  saveData("categories", updatedCategories);
  
  // Also remove any transactions with this category
  const transactions = getTransactions();
  const updatedTransactions = transactions.filter(t => t.categoryId !== id);
  saveData("transactions", updatedTransactions);
};

// Get transaction summary
export const getTransactionSummary = () => {
  const transactions = getTransactions();
  
  const totalIncome = transactions
    .filter(t => t.type === "income")
    .reduce((total, t) => total + t.amount, 0);
    
  const totalExpense = transactions
    .filter(t => t.type === "expense")
    .reduce((total, t) => total + t.amount, 0);
    
  const balance = totalIncome - totalExpense;
  
  return {
    totalIncome,
    totalExpense,
    balance
  };
};

// Get category summary
export const getCategorySummary = (type: "income" | "expense") => {
  const transactions = getTransactions();
  const categories = getCategories();
  
  // Filter by transaction type
  const filteredTransactions = transactions.filter(t => t.type === type);
  
  // Group by category
  const categorySummary = categories
    .filter(c => c.type === type || c.type === "both")
    .map(category => {
      const categoryTransactions = filteredTransactions.filter(
        t => t.categoryId === category.id
      );
      
      const total = categoryTransactions.reduce(
        (sum, t) => sum + t.amount, 
        0
      );
      
      return {
        category,
        total,
        count: categoryTransactions.length
      };
    })
    .filter(summary => summary.count > 0)
    .sort((a, b) => b.total - a.total);
    
  return categorySummary;
};
